modules/support_ticket/src/Entity/SupportTicket.php

```diff
@@ -0,0 +1,468 @@
+ *   config_prefix = "form",
+ *   entity_keys = {
+ *     "id" = "stid",
+ *     "revision" = "vid",
+ *     "bundle" = "type",
+ *     "label" = "title",
+ *     "langcode" = "langcode",
+ *     "uuid" = "uuid",
+ *     "status" = "status",
+ *     "uid" = "uid",
+ *   },
+ *   bundle_entity_type = "support_ticket_type",
+ *   field_ui_base_route = "entity.support_ticket_type.edit_form",
+ *   links = {
+ *     "canonical" = "/support_ticket/{support_ticket}",
+ *     "delete-form" = "/support_ticket/{support_ticket}/delete",
+ *     "edit-form" = "/support_ticket/{support_ticket}/edit",
+ *     "version-history" = "/support_ticket/{support_ticket}/revisions",
+ *     "revision" = "/support_ticket/{support_ticket}/revisions/{support_ticket_revision}/view",
+ *   }
+ * )
+ */
+class SupportTicket extends ContentEntityBase implements SupportTicketInterface {
+
+  use EntityChangedTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function preSave(EntityStorageInterface $storage) {
+    parent::preSave($storage);
+
+    // If no owner has been set explicitly, make the current user the owner.
+    if (!$this->getOwner()) {
+      $this->setOwnerId(\Drupal::currentUser()->id());
+    }
+    // If no revision author has been set explicitly, make the support ticket owner
+    // the revision author.
+    if (!$this->getRevisionAuthor()) {
+      $this->setRevisionAuthorId($this->getOwnerId());
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function preSaveRevision(EntityStorageInterface $storage, \stdClass $record) {
+    parent::preSaveRevision($storage, $record);
+
+    if (!$this->isNewRevision() && isset($this->original) && (!isset($record->revision_log) || $record->revision_log === '')) {
+      // If we are updating an existing support ticket without adding a new revision,
+      // we need to make sure $entity->revision_log is reset whenever it is empty.
+      // Therefore, this code allows us to avoid clobbering an existing log
+      // entry with an empty one.
+      $record->revision_log = $this->original->revision_log->value;
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function postSave(EntityStorageInterface $storage, $update = TRUE) {
+    parent::postSave($storage, $update);
+
+    // Reindex the support_ticket when it is updated. The support_ticket is
+    // automatically indexed when it is added, simply by being added to the
+    // support_ticket table.
+    if ($update) {
+      // @todo
+      // support_ticket_reindex_support_ticket_search($this->id());
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function preDelete(EntityStorageInterface $storage, array $entities) {
+    parent::preDelete($storage, $entities);
+
+    // Ensure that all support_tickets deleted are removed from the search index.
+    if (\Drupal::moduleHandler()->moduleExists('search')) {
+      foreach ($entities as $entity) {
+        search_index_clear('support_ticket_search', $entity->stid->value);
```

Comment:
Not an issue for here, but I wonder a bit whether search indexing could re-use cache metadata.